### PR TITLE
Checkpoint the RNG the curriculum sampler actually draws from

### LINKED ISSUE #8405: [BUG] Curriculum data sampler checkpoints the global numpy RNG, not its own, so resume replays the sampling stream
**Describe the bug**

`DeepSpeedDataSampler` draws from its own generator, `self.np_rng = np.random.default_rng(seed)`, but `state_dict()` stores `np.random.get_state()` and `load_state_dict()` restores it. That is the global legacy `RandomState`, not the generator the sampler uses, so the saved value is the same whether the sampler has drawn nothing or a thousand batches.

The effect shows up on resume. `__init__` seeds `self.np_rng` fresh, `load_state_dict` never touches it, and the sampler goes back to the top of its own stream: same cluster mix per step, same shuffles out of `get_new_cluster` and `reshuffle_clusters`. Restoring the global state also moves whatever else in the process is drawing from `np.random`.

**To Reproduce**

```python
from deepspeed.runtime.data_pipeline.config import get_data_efficiency_config
from deepspeed.runtime.data_pipeline.data_sampling.data_sampler import DeepSpeedDataSampler

metric = {"index_to_sample_path": "dummy", "index_to_metric_path": "dummy", "difficulty_type": "value",
          "clustering_type": "single_cluster", "min_difficulty": 8, "max_difficulty": 80,
          "schedule_type": "fixed_linear",
          "schedule_config": {"total_curriculum_step": 100, "difficulty_step": 8}}
config = get_data_efficiency_config({"data_efficiency": {"enabled": True, "seed": 1234, "data_sampling": {
    "enabled": True, "curriculum_learning": {"enabled": True, "data_cluster_path": "/tmp/clusters",
                                             "curriculum_metrics": {"dummy": metric}}}}})

def sampler():
    s = DeepSpeedDataSampler(config, 100, 8, 0, 1, None, 1, global_rank=0)
    s.data_clusters, s.data_cluster_sizes = [None] * 4, [10, 20, 30, 40]
    return s

saved = sampler()
before = saved.state_dict()["np_rng_state"]
first = [saved.sample_from_clusters().tolist() for _ in range(3)]
print("saved rng state changed by 3 draws:", repr(saved.state_dict()["np_rng_state"]) != repr(before))

resumed = sampler()
resumed.load_state_dict(saved.state_dict())
print("resumed draw 1:", resumed.sample_from_clusters().tolist(), " original draw 1:", first[0])
```

On master at `493dafa`, python 3.12.14, torch 2.14.0+cpu:

```
saved rng state changed by 3 draws: False
resumed draw 1: [0, 3, 3, 2]  original draw 1: [0, 3, 3, 2]
```

**Expected behavior**

A resumed run continues the sampling stream instead of replaying it, and loading a sampler state leaves the global numpy RNG alone.

**Additional context**

Both lines go back to #2585, which added the library. I found this reading the code, not from a training run. PR on the way.

### PR DESCRIPTION
**Pull request by alanhuangyoo**
Fixes #8405, reported by @ebarkhordar.

## Problem

Every draw in `DeepSpeedDataSampler` goes through `self.np_rng`, a `np.random.default_rng` seeded in `__init__`:

```python
self.np_rng = np.random.default_rng(self.data_efficiency_config[DATA_EFFICIENCY_SEED])   # :64
...
self.np_rng.shuffle(new_cluster)                                                          # :217  get_new_cluster
self.np_rng.choice(num_clusters, self.global_batch_size, replace=True, p=weights)         # :236  sample_from_clusters
self.np_rng.shuffle(cluster)                                                              # :246  reshuffle_clusters
self.np_rng.shuffle(batch)                                                                # :289
```

but the checkpoint stored and restored something else entirely:

```python
CURRICULUM_LEARNING_NP_RNG_STATE: np.random.get_state()      # :324  state_dict
np.random.set_state(state_dict[CURRICULUM_LEARNING_NP_RNG_STATE])   # :334  load_state_dict
```

That is the process-wide legacy `RandomState`, which this sampler never touches. So the saved value reads the same whether the sampler has drawn nothing or a thousand batches, and on resume `__init__` re-seeds `self.np_rng` while `load_state_dict` leaves it alone — the sampler restarts its own stream, reproducing the same cluster mix per step and the same shuffles it produced the first time. Restoring the global state also moved whatever else in the process draws from `np.random`.

## Solution

Save and restore `self.np_rng.bit_generator.state`.

A checkpoint written before this holds the legacy tuple, and there is no sampler stream recorded in it to restore. `_load_np_rng_state` recognises the shape, warns once on rank 0, and leaves the sampler on its configured seed — which is where those checkpoints already resumed — rather than refusing to load.

## Verification

`tests/unit/runtime/data_pipeline/test_curriculum_sampler_rng.py`, four cases:

| test | on master | here |
| --- | --- | --- |
| the saved state moves as the sampler draws | ❌ identical before and after 3 draws | ✅ |
| resume continues the stream rather than replaying it | ❌ `[[0, 3, 3, 2], ...] != [[2, 1, 3, 2], ...]` | ✅ |
| loading leaves the global numpy RNG alone | ✅ | ✅ |
| a legacy checkpoint still loads | ✅ | ✅ |

The resume test goes through `load_state_dict` rather than poking the generator, so it fails on master for the behaviour rather than on a type error, and it asserts that a sampler on a fresh seed does *not* already agree — otherwise it would prove nothing.

```
1×H20:  4 passed   (this branch)
        2 failed, 2 passed   (master, same test file)
```

`pre-commit run` passes on both files.

### CODE REVIEW

**Comment by alanhuangyoo** [comment]
Closing this one. #8489 lists the Data Efficiency library as a deprecation candidate for lack of bandwidth, so a checkpointing fix to its sampler probably isn't worth anyone's review time now. If Data Efficiency ends up staying, I'm happy to reopen. The change is a single file plus its test.
